Create Downloadable Spreadsheet using SheetJS

IshaGupta18 opened the pull request:
So here's the first of the export options I have implemented: The user can download his uploaded CSV data as an excel spreadsheet. 
![sheetjs](https://user-images.githubusercontent.com/40794215/59300338-91ce7100-8cac-11e9-8259-036262f33ca5.gif)
@jywarren what do you think of this? This is a one-click export option! #17 

@namangupta01 @gauravano @IgorWilbert @Souravirus what do you think?

jywarren commented:
Wow, this is very cool! What do you think, ready to merge?

jywarren commented on dist/PublicLab.Grapher.js:
```diff
@@ -235,10 +275,108 @@ function () {
+      this.csvFile = mat;
+
+      document.getElementById(this.uploadButtonId).onclick = function (e) {
+        console.log("i am uploading");
```
ah i see not quite perhaps but once the console.log statements are removed? Thanks!

jywarren commented:
Isha, what do you think about this idea - could be discussed in a new issue - @publiclab-mimi was asking about saving data from https://webjack.io/examples/ -- to make it easy for students in a classroom to record sensor data from an Arduino, and visualize it. What about adding a button to WebJack that says "view this data in a graph" that opens an instance of Simple Data Grapher?

IshaGupta18 commented:
That would be a pretty awesome idea. However, could you guide me a little
on how we could go about implementing this? This would be great to work on!

Also, yes this is ready for merge, it has a couple of debugging console.log
statements and I think one small little problem which don't affect the
SheetJS function, so if you could merge this, I could fix the small issues
in a follow up and Naman can also view them there. What do you think?

On Wed, Jun 12, 2019, 4:32 AM Jeffrey Warren <notifications@github.com>
wrote:

> Isha, what do you think about this idea - could be discussed in a new
> issue - @publiclab-mimi <https://github.com/publiclab-mimi> was asking
> about saving data from https://webjack.io/examples/ -- to make it easy
> for students in a classroom to record sensor data from an Arduino, and
> visualize it. What about adding a button to WebJack that says "view this
> data in a graph" that opens an instance of Simple Data Grapher?
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/publiclab/simple-data-grapher/pull/43?email_source=notifications&email_token=AJXHQZ5S5PBXTMSMFTFAWLTP2AVIRA5CNFSM4HXCT4TKYY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGODXOX6FQ#issuecomment-501055254>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AJXHQZ2MLNMYTVZDX2DBLDLP2AVIRANCNFSM4HXCT4TA>
> .
>

IshaGupta18 commented on src/View.js:
```diff
@@ -41,10 +44,90 @@ class View{
+        console.log(vall);
+        console.log("hurray!!");
+    }
+    handleFileSelectremote(val){
```
#45 This is the code snippet

jywarren replied:
This is really nicely structured code! Easy to read!

IshaGupta18 replied:
Thank you!

jywarren commented on dist/PublicLab.Grapher.js:
```diff
@@ -235,10 +275,108 @@ function () {
+      document.getElementById(this.uploadButtonId).onclick = function (e) {
+        console.log("i am uploading");
+        console.log(this);
+        this.csvParser = new _CsvParser.CsvParser(this.csvFile, this.elementId, "csvstring");
```
Ah, i see - is this the error you mentioned in #44? An inline comment like this can help folks quickly sync up with you. 👍

jywarren replied:
This is a great candidate for unit testing because you're instantiating a whole object here, that could be done in a more isolated situation in a unit test. Would be cool!

jywarren replied:
Copying in error from #44: `Uncaught ReferenceError: _CsvParser is not defined`

jywarren replied:
It looks like you changed the name to `_CsvParser2` on this line? https://github.com/publiclab/simple-data-grapher/pull/43/files#diff-879bf0cd07df9ada5c670f6eb2195ab1R246

IshaGupta18 replied:
The dist files are generated through browserify and babel, so I didn't change the name here, it was automatically changed when I wrote code in the View.js and CsvParser.js

jywarren commented on dist/transpiled_code/View.js:
```diff
@@ -5,7 +5,7 @@ Object.defineProperty(exports, "__esModule", {
 exports.View = void 0;
 
-var _CsvParser = require("./CsvParser");
+var _CsvParser2 = require("./CsvParser");
```
Yes, here you could use a unit test to check out the functions of CsvParser, which would be good to give us confidence that it works as intended even before integrating it into the rest of the system...!

IshaGupta18 replied:
Yes it can be done! Once this bug is resolved, I think we can write the tests!

rexagod commented on dist/transpiled_code/View.js:
```diff
@@ -37,10 +37,108 @@ function () {
+      document.getElementById(this.uploadButtonId).onclick = function (e) {
+        console.log("i am uploading");
+        console.log(this);
+        this.csvParser = new _CsvParser.CsvParser(this.csvFile, this.elementId, "csvstring");
```
So I guess the problem here is that `new _CsvParser` isn't getting changed to `new _CsvParser2`? If so, would you mind doing a rename of this object to anything other than its original returned variable name?

IshaGupta18 commented:
Hey thanks a lot! I just fixed this bug. Phewww.... But I will have to test
it a little before I can say for sure. I'll back to you if I face another
problem with this. Thank you!

On Thu, Jun 13, 2019, 11:36 PM Pranshu Srivastava <notifications@github.com>
wrote:

> *@rexagod* requested changes on this pull request.
> ------------------------------
>
> In dist/transpiled_code/View.js
> <https://github.com/publiclab/simple-data-grapher/pull/43#discussion_r293509101>
> :
>
> > +          continue;
> +        }
> +
> +        var dataHash = Papa.parse(csv_string[i], {
> +          dynamicTyping: true,
> +          comments: true
> +        });
> +        mat[i] = dataHash['data'][0];
> +      }
> +
> +      this.csvFile = mat;
> +
> +      document.getElementById(this.uploadButtonId).onclick = function (e) {
> +        console.log("i am uploading");
> +        console.log(this);
> +        this.csvParser = new _CsvParser.CsvParser(this.csvFile, this.elementId, "csvstring");
>
> So I guess the problem here is that new _CsvParser isn't getting changed
> to new _CsvParser2? If so, would you mind doing a rename of this object
> to anything other than its original returned variable name?
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/publiclab/simple-data-grapher/pull/43?email_source=notifications&email_token=AJXHQZ4BLLJZ7EKULFN76EDP2KEBRA5CNFSM4HXCT4TKYY3PNVWWK3TUL52HS4DFWFIHK3DMKJSXC5LFON2FEZLWNFSXPKTDN5WW2ZLOORPWSZGOB3PVE2I#pullrequestreview-249516649>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AJXHQZ4ZWZ7DF36QGSKRJDTP2KEBRANCNFSM4HXCT4TA>
> .
>

rexagod commented:
Sure, no problem!

IshaGupta18 commented:
Hey @jywarren I have finally resolved #45 with @rexagod 's help and #44 is also resolved now. Apart from this, I have implemented all the 4 parsing functions, namely Remote, Local, Google Sheet and String file upload. I have also implemented the "save as image" function for every graph and save as spreadsheet for the data imported. Besides this, I have fixed some other small bugs and tried to modularize the code further. I know this is a bug PR, but its ready for merging. What do you think?

jywarren commented:
Wow, amazing! Let me read through -- has @namangupta01 taken a look over it as well?

jywarren commented on dist/transpiled_code/View.js:
```diff
@@ -33,14 +31,129 @@ function () {
+  }, {
+    key: "handleFileSelectstring",
+    value: function handleFileSelectstring(val) {
+      console.log("i am at csv string handler");
```
Oops watch the comments here! Maybe remove? Thanks!

IshaGupta18 commented:
I think he will be okay with it! He is been keeping a little busy these
days.

On Tue, Jun 18, 2019, 7:38 PM Jeffrey Warren <notifications@github.com>
wrote:

> Wow, amazing! Let me read through -- has @namangupta01
> <https://github.com/namangupta01> taken a look over it as well?
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/publiclab/simple-data-grapher/pull/43?email_source=notifications&email_token=AJXHQZ7IQOI7QLNBKSKRW4TP3DT5XA5CNFSM4HXCT4TKYY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGODX6X3EA#issuecomment-503152016>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AJXHQZZV73H2AMOBF4TX2P3P3DT5XANCNFSM4HXCT4TA>
> .
>

IshaGupta18 commented:
I have deliberately left them for now, will be needing it in debugging.
Once we are a little more confident with our code, I'll get rid of them.
Does that sound right?

On Tue, Jun 18, 2019, 7:39 PM Jeffrey Warren <notifications@github.com>
wrote:

> *@jywarren* commented on this pull request.
> ------------------------------
>
> In dist/transpiled_code/View.js
> <https://github.com/publiclab/simple-data-grapher/pull/43#discussion_r294842598>
> :
>
> >          };
>        }
>      }
> +  }, {
> +    key: "handleFileSelectstring",
> +    value: function handleFileSelectstring(val) {
> +      console.log("i am at csv string handler");
>
> Oops watch the comments here! Maybe remove? Thanks!
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/publiclab/simple-data-grapher/pull/43?email_source=notifications&email_token=AJXHQZ5DYDF6P5ORW2IMBKLP3DUCNA5CNFSM4HXCT4TKYY3PNVWWK3TUL52HS4DFWFIHK3DMKJSXC5LFON2FEZLWNFSXPKTDN5WW2ZLOORPWSZGOB335OYI#pullrequestreview-251123553>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AJXHQZ72BNL3OGPPB23SNXDP3DUCNANCNFSM4HXCT4TA>
> .
>

jywarren commented on src/View.js:
```diff
@@ -359,13 +493,31 @@ class View{
             console.log("i am here23");
             this.handleFileSelectlocal(e);
         });
+        $("#"+this.csvStringUploadId).change(()=>{
```
Starting to get a bit long in this file; is there any way to further break it up into separate functions/files in a follow-up PR?

IshaGupta18 replied:
Yes we can definitely try and do that!

IshaGupta18 commented:
Yayy, thank you!!

On Tue, Jun 18, 2019, 7:41 PM Jeffrey Warren <notifications@github.com>
wrote:

> *@jywarren* approved this pull request.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/publiclab/simple-data-grapher/pull/43?email_source=notifications&email_token=AJXHQZ6DVPN2L4BCFDWAR6DP3DUKBA5CNFSM4HXCT4TKYY3PNVWWK3TUL52HS4DFWFIHK3DMKJSXC5LFON2FEZLWNFSXPKTDN5WW2ZLOORPWSZGOB335ZSI#pullrequestreview-251124937>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AJXHQZ3TUWZ7CFMAIKZDQN3P3DUKBANCNFSM4HXCT4TA>
> .
>

jywarren commented:
:-) Hooray!
